File: src/main/java/org/tindalos/principle/infrastructure/service/jdepend/JDependBasedPackage.java
package org.tindalos.principle.infrastructure.service.jdepend;

import java.util.Collection;
import java.util.Set;

import jdepend.framework.JavaPackage;

import org.tindalos.principle.domain.core.Metrics;
import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.core.PackageReference;

import com.google.common.collect.Sets;

public class JDependBasedPackage extends Package {
	
    private Set<PackageReference> ownPackageReferences;
    private final Metrics metrics;
	private final boolean isUnreferred;

	public JDependBasedPackage(JavaPackage javaPackage, String basePackage, Metrics metrics) {
        super(javaPackage.getName());
        this.metrics = metrics;
        this.isUnreferred = metrics.getAfferentCoupling() == 0;
        this.ownPackageReferences =  createOwnPackageReferences(javaPackage, basePackage);
    }
	
	@Override
	public Set<PackageReference> getOwnPackageReferences() {
		return ownPackageReferences;
	}

    @SuppressWarnings("unchecked")
    protected Set<PackageReference> createOwnPackageReferences(JavaPackage theJavaPackage, String basePackage) {

        Set<PackageReference> packages = Sets.newHashSet();

        Collection<JavaPackage> efferents = theJavaPackage.getEfferents();
        for (JavaPackage javaPackage : efferents) {
            if (javaPackage.getName().startsWith(basePackage)) {
                packages.add(new PackageReference(javaPackage.getName()));
            }
        }

        return packages;
    }

    @Override
    public Metrics getMetrics() {
        return metrics;
    }

	@Override
	public boolean isUnreferred() {
		return this.isUnreferred;
	}

}

File: src/main/scala/org/tindalos/principle/infrastructure/service/jdepend/JDependBasedPackage.scala
package org.tindalos.principle.infrastructure.service.jdepend

import java.util.Collection

import jdepend.framework.JavaPackage
import org.tindalos.principle.domain.core.{Metrics, Package, PackageReference}

import scala.collection.JavaConversions._

class JDependBasedPackage(val javaPackage: JavaPackage, val basePackage: String, val metrics: Metrics)
  extends Package(javaPackage.getName()) {

  private val validExternalEfferents = Set("java", "scala")

  override def getMetrics() = metrics

  override def isUnreferred() = metrics.afferentCoupling == 0

  override def getOwnPackageReferences() = ownPackageReferences

  override def getOwnExternalPackageReferences(): Set[PackageReference] = ownExternalPackageReferences

  private lazy val ownPackageReferences =
    javaPackage.getEfferents().asInstanceOf[Collection[JavaPackage]]
      .filter(_.getName().startsWith(basePackage))
      .map(x ⇒ new PackageReference(x.getName()))
      .toSet

  private lazy val ownExternalPackageReferences =

    javaPackage.getEfferents().asInstanceOf[Collection[JavaPackage]]
      .filter(p ⇒ !p.getName().startsWith(basePackage) && isNotValidExternalEfferent(p))
      .map(x ⇒ new PackageReference(x.getName()))
      .toSet

  def isNotValidExternalEfferent(p: JavaPackage) = !validExternalEfferents.exists(e ⇒ p.getName().startsWith(e))

}